cmd/pruner.go

```diff
@@ -97,538 +191,75 @@ func pruneAppState(home string) error {
-			"authz",    // authzkeeper.StoreKey,
-		)
+	}
 
-		for key, value := range fetchhubKeys {
-			keys[key] = value
-		}
-	} else if app == "persistent" {
-		persistentKeys := types.NewKVStoreKeys(
-			"halving",  // halving.StoreKey,
-			"authz",    // sdkAuthzKeeper.StoreKey,
-			"feegrant", // feegrant.StoreKey,
-		)
+	extraKeys := types.NewKVStoreKeys(modules...)
 
-		for key, value := range persistentKeys {
-			keys[key] = value
-		}
-	} else if app == "cryptoorgchain" {
-		cryptoorgchainKeys := types.NewKVStoreKeys(
-			"feegrant",  // feegrant.StoreKey,
-			"authz",     // authzkeeper.StoreKey,
-			"chainmain", // chainmaintypes.StoreKey,
-			"supply",    // supplytypes.StoreKey,
-			"nft",       // nfttypes.StoreKey,
-		)
+	for key, value := range extraKeys {
+		keys[key] = value
+	}
 
-		for key, value := range cryptoorgchainKeys {
-			keys[key] = value
-		}
-	} else if app == "irisnet" {
-		irisnetKeys := types.NewKVStoreKeys(
-			"guardian", // guardiantypes.StoreKey,
-			"token",    // tokentypes.StoreKey,
-			"nft",      // nfttypes.StoreKey,
-			"htlc",     // htlctypes.StoreKey,
-			"record",   // recordtypes.StoreKey,
-			"coinswap", // coinswaptypes.StoreKey,
-			"service",  // servicetypes.StoreKey,
-			"oracle",   // oracletypes.StoreKey,
-			"random",   // randomtypes.StoreKey,
-			"farm",     // farmtypes.StoreKey,
-			"feegrant", // feegrant.StoreKey,
-			"tibc",     // tibchost.StoreKey,
-			"NFT",      // tibcnfttypes.StoreKey,
-			"MT",       // tibcmttypes.StoreKey,
-			"mt",       // mttypes.StoreKey,
-		)
+	wg := sync.WaitGroup{}
+	var prune_err error
 
-		for key, value := range irisnetKeys {
-			keys[key] = value
-		}
-	} else if app == "axelar" {
-		axelarKeys := types.NewKVStoreKeys(
-			"feegrant",   // feegrant.StoreKey,
-			"vote",       // voteTypes.StoreKey,
-			"evm",        // evmTypes.StoreKey,
-			"snapshot",   // snapTypes.StoreKey,
-			"tss",        // tssTypes.StoreKey,
-			"nexus",      // nexusTypes.StoreKey,
-			"axelarnet",  // axelarnetTypes.StoreKey,
-			"reward",     // rewardTypes.StoreKey,
-			"permission", // permissionTypes.StoreKey,
-		)
+	guard := make(chan struct{}, parallel)
+	for _, value := range keys {
+		guard <- struct{}{}
+		wg.Add(1)
+		go func(value *types.KVStoreKey) {
+			tmp_err := func(value *types.KVStoreKey) error {
+				// TODO: cleanup app state
+				appStore := rootmulti.NewStore(appDB)
+				appStore.MountStoreWithDB(value, sdk.StoreTypeIAVL, nil)
+				err = appStore.LoadLatestVersion()
+				if err != nil {
+					return err
+				}
 
-		for key, value := range axelarKeys {
-			keys[key] = value
-		}
-	} else if app == "umee" {
-		umeeKeys := types.NewKVStoreKeys(
-			"feegrant", // feegrant.StoreKey,
-			"authz",    // authzkeeper.StoreKey,
-			"gravity",  // gravitytypes.StoreKey,
-		)
+				versions := appStore.GetAllVersions()
+				if int(keepVersions) >= len(versions) {
+					return nil
+				}
+				versions = versions[:len(versions)-int(keepVersions)]
```

Comment:
I think we can't use length. 

for example, we have versions like this.
--> 20,40,60,80,100,101,102,103,104,105 (recent: 5, every: 20)
We want to prune with (recent: 20, every: 0). 

The final result should be 100,101,102,103,104,105.
If we use length, it won't prune anything because it will keep the last 20 available versions.

Reply:
Wow Genius